### PR TITLE
Add a powermock alternative

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: -
- Related Pull Requests: -

* * *

## What
This pull request adds the new class BaseTest to the repository that should be extended by all test classes.

## Why
This PR is needed because everything is encapsulated as much as possible, while the tests do need access even to the strongly encapsulated classes.

## How
The new class BaseTest uses reflection to make it easy for the programmer to invoke private constructors / methods and access private fields.

## Alternative implementation
I have tried Powermock by using electro, but I couldn't get it working. I suspect it's because we're compiling using the most recent Android build tools.

## Notes
-

### CODE REVIEW

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package com.nervousfish.nervousfish;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * The superclass of all tests
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: This is no longer accurate

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ */
+public final class BaseTest {
+
+    /**
+     * Prevents instantiation
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: "Simple constructor for the BaseTest class"

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    private BaseTest() {
+        // Prevents instantiation
+    }
+
+    /**
+     * Creates a new instance of a class that has an inaccessible constructor (eg private)
+     * Note that there should be a single constructor in the class
+     *
+     * @param clazz The class that should be instantiated
+     * @param args  The arguments that should be passed to the constructor
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Are we left aligning `@param` descriptions or not?

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Prevents instantiation
+     */
+    private BaseTest() {
+        // Prevents instantiation
+    }
+
+    /**
+     * Creates a new instance of a class that has an inaccessible constructor (eg private)
+     * Note that there should be a single constructor in the class
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Should end with a period and "eg" should be "e.g."

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                return constructor.newInstance(args);
+            }
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            e.printStackTrace();
+        }
+        throw new RuntimeException();
+    }
+
+    /**
+     * Executes an inaccessible method on an object (eg private method)
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Should end with a period and "eg" should be "e.g."

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                }
+            }
+        } catch (InvocationTargetException | IllegalAccessException e) {
+            e.printStackTrace();
+        }
+        throw new RuntimeException();
+    }
+
+    /**
+     * Sets the value of an inaccessible field on an object (eg private field)
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Should end with a period and "eg" should be "e.g."

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            field.setAccessible(true);
+            field.set(object, value);
+        } catch (IllegalAccessException | NoSuchFieldException e) {
+            e.printStackTrace();
+        }
+        throw new RuntimeException();
+    }
+
+    /**
+     * Gets the value of an inaccessible field on an object (eg private field)
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Should end with a period and "eg" should be "e.g."

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java (hunk 3 above)
Comment (jverbraeken) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Don't know, I just press Ctrl-Alt-Shift-L and then Intellij aligns everything to the official Java standard

**Comment by jverbraeken** [comment]
@ericcornelissen Done!

File: app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java [hunk 8]
```diff
@@ -13,17 +21,27 @@
  */
 @SuppressWarnings("PMD")
 public class ExampleUnitTest {
+    ISample sample = mock(ISample.class);
+
     @Test
     public void addition_isCorrect() throws Exception {
```
Comment (TheBonheurs) [inline app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java]: What is the point of this test?

File: app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java [hunk 9]
```diff
@@ -13,17 +21,27 @@
...
         assertEquals(4, 2 + 2);
     }
 
+    @Test
+    public void bar() {
+        when(sample.foo()).thenReturn(5);
+        assertEquals(new SampleClass(sample).get(), 5);
+        Constants constants = (Constants) accessConstructor(Constants.class, mock(IServiceLocatorCreator.class));
```
Comment (TheBonheurs) [inline app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java]: Why declare a constants here, if it's an example why not try doing something with it?

Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java]: It is an example of `accessConstructor`, if no error is thrown it should be working

File: app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java (hunk 8 above)
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java]: It was there by default, but it doesn't matter as this class will be removed once the first real unittests are added to the repo

**Pull request merged**